diff -r -N -u openeclair-android_packages_apps_Mms-4e5c488/res/layout/custom_user_agent_dialog.xml openeclair-android_packages_apps_Mms-4e5c488-ua/res/layout/custom_user_agent_dialog.xml
--- openeclair-android_packages_apps_Mms-4e5c488/res/layout/custom_user_agent_dialog.xml	1970-01-01 01:00:00.000000000 +0100
+++ openeclair-android_packages_apps_Mms-4e5c488-ua/res/layout/custom_user_agent_dialog.xml	2010-03-08 15:24:13.659076856 +0000
@@ -0,0 +1,33 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinearLayout
+  android:layout_height="wrap_content"
+  xmlns:android="http://schemas.android.com/apk/res/android"
+  android:layout_width="fill_parent"
+  android:orientation="vertical"
+  android:paddingLeft="5sp"
+  android:paddingRight="5sp"
+  android:id="@id/CustomUserAgentLinearLayout">
+  <ScrollView
+    android:layout_height="wrap_content"
+    android:layout_width="fill_parent"
+    android:id="@id/CustomUserAgentScrollView">
+    <LinearLayout
+      android:layout_height="wrap_content"
+      android:orientation="vertical"
+      android:layout_width="fill_parent"
+      android:id="@id/ScrollViewLinearLayout"
+      android:paddingLeft="5sp"
+      android:paddingRight="5sp"
+      android:paddingBottom="2sp"
+      android:paddingTop="2sp">
+      <EditText
+        android:id="@id/CustomUserAgentEditText"
+        android:layout_width="fill_parent"
+        android:layout_height="wrap_content"
+        android:singleLine="true"
+        android:maxLength="1000"
+        android:padding="5sp" />
+    </LinearLayout>
+  </ScrollView>
+</LinearLayout>
+
diff -r -N -u openeclair-android_packages_apps_Mms-4e5c488/res/values/arrays.xml openeclair-android_packages_apps_Mms-4e5c488-ua/res/values/arrays.xml
--- openeclair-android_packages_apps_Mms-4e5c488/res/values/arrays.xml	2010-03-07 21:27:58.000000000 +0000
+++ openeclair-android_packages_apps_Mms-4e5c488-ua/res/values/arrays.xml	2010-03-08 15:26:08.887075605 +0000
@@ -97,23 +97,24 @@
       <item>0,800,200,800,200,800</item>
    </string-array>
 
-	<string-array name="pref_mms_user_agent_entries">
-	    <item>Default Android</item>
-	    <item>Nokia N95</item>
-	    <item>iPhone (3.0)</item>
-	    <item>BlackBerry 8800</item>
-	    <item>Sony Ericsson W800i</item>
-	    <item>SoftBank X01T</item>
-	    <!-- <item>@string/pref_mms_custom_option</item> -->
-	  </string-array>
+   <string-array name="pref_mms_user_agent_entries">
+    <item>Default Android</item>
+    <item>Nokia N95</item>
+    <item>iPhone (3.0)</item>
+    <item>BlackBerry 8800</item>
+    <item>Sony Ericsson W800i</item>
+    <item>SoftBank X01T</item>
+    <item>@string/pref_mms_custom_option</item>
+  </string-array>
+ 
+  <string-array name="pref_mms_user_agent_values">
+    <item>@string/pref_key_mms_user_agent_default</item>
+    <item>NokiaN95/11.0.026; Series60/3.1 Profile/MIDP-2.0 Configuration/CLDC-1.1</item>
+    <item>iPhoneOS/3.0 (7A341)</item>
+    <item>BlackBerry8800/4.2.1 Profile/MIDP-2.0 Configuration/CLDC-1.1 VendorID/100</item>
+    <item>SonyEricssonW800i/R1L</item>
+    <item>SoftBank/1.0/X01T/SCJ001</item>
+    <item>custom</item>
+  </string-array>
 
-	  <string-array name="pref_mms_user_agent_values">
-	    <!-- <item>@string/pref_key_mms_user_agent_default</item> -->
-	    <item>NokiaN95/11.0.026; Series60/3.1 Profile/MIDP-2.0 Configuration/CLDC-1.1</item>
-	    <item>iPhoneOS/3.0 (7A341)</item>
-	    <item>BlackBerry8800/4.2.1 Profile/MIDP-2.0 Configuration/CLDC-1.1 VendorID/100</item>
-	    <item>SonyEricssonW800i/R1L</item>
-	    <item>SoftBank/1.0/X01T/SCJ001</item>
-	    <item>custom</item>
-	  </string-array>
 </resources>
diff -r -N -u openeclair-android_packages_apps_Mms-4e5c488/res/values/strings.xml openeclair-android_packages_apps_Mms-4e5c488-ua/res/values/strings.xml
--- openeclair-android_packages_apps_Mms-4e5c488/res/values/strings.xml	2010-03-07 21:27:58.000000000 +0000
+++ openeclair-android_packages_apps_Mms-4e5c488-ua/res/values/strings.xml	2010-03-08 15:27:36.683075781 +0000
@@ -369,6 +369,12 @@
     <!-- Settings item description for boolean MMS auto-retrieve while roaming option -->
     <string name="pref_summary_mms_retrieval_during_roaming">Automatically retrieve messages while roaming</string>
 
+    <string name="pref_mms_custom_option">Custom..</string>
+    <string name="pref_key_mms_user_agent_default">Android-Mms/2.0</string>
+    <string name="pref_title_mms_user_agent">Custom User-Agent</string>
+    <string name="pref_summary_mms_user_agent">Spoof phone as a different model for MMS</string>
+    <string name="pref_mms_user_agent_custom_set">Custom User Agent set to:\n</string> 
+
     <!-- Label prepended to recipient's name in Undelivered Messages screen -->
     <string name="to_label">"To "</string>
 
diff -r -N -u openeclair-android_packages_apps_Mms-4e5c488/res/xml/preferences.xml openeclair-android_packages_apps_Mms-4e5c488-ua/res/xml/preferences.xml
--- openeclair-android_packages_apps_Mms-4e5c488/res/xml/preferences.xml	2010-03-07 21:27:58.000000000 +0000
+++ openeclair-android_packages_apps_Mms-4e5c488-ua/res/xml/preferences.xml	2010-03-08 15:28:33.395074872 +0000
@@ -139,5 +139,13 @@
                         android:title="@string/pref_title_mms_send_on_enter"
                         android:summary="@string/pref_summary_mms_send_on_enter"
                         android:defaultValue="true" />
+    <com.android.mms.preferences.UserAgentListPreference
+                        android:key="pref_key_mms_user_agent"
+                        android:defaultValue="@string/pref_key_mms_user_agent_default"
+                        android:title="@string/pref_title_mms_user_agent"
+                        android:summary="@string/pref_summary_mms_user_agent"
+                        android:entries="@array/pref_mms_user_agent_entries"
+                        android:entryValues="@array/pref_mms_user_agent_values"
+                        android:dialogTitle="@string/pref_title_mms_user_agent" /> 
   </PreferenceCategory>
 </PreferenceScreen>
diff -r -N -u openeclair-android_packages_apps_Mms-4e5c488/src/com/android/mms/preferences/UserAgentListPreference.java openeclair-android_packages_apps_Mms-4e5c488-ua/src/com/android/mms/preferences/UserAgentListPreference.java
--- openeclair-android_packages_apps_Mms-4e5c488/src/com/android/mms/preferences/UserAgentListPreference.java	1970-01-01 01:00:00.000000000 +0100
+++ openeclair-android_packages_apps_Mms-4e5c488-ua/src/com/android/mms/preferences/UserAgentListPreference.java	2010-03-08 15:30:38.519075823 +0000
@@ -0,0 +1,126 @@
+package com.android.mms.preferences;
+ 
+import com.android.mms.R;
+import com.android.mms.ui.MessagingPreferenceActivity;
+ 
+import android.app.AlertDialog;
+import android.content.Context;
+import android.content.DialogInterface;
+import android.content.SharedPreferences;
+import android.content.DialogInterface.OnCancelListener;
+ 
+import android.util.AttributeSet;
+import android.view.LayoutInflater;
+import android.view.View;
+import android.os.Parcelable;
+import android.preference.ListPreference;
+import android.preference.PreferenceManager;
+import android.widget.SeekBar;
+import android.widget.EditText;
+import android.widget.Toast;
+ 
+public class UserAgentListPreference extends ListPreference {
+    private Context mContext;
+    private String mUserAgent;
+    private boolean mDialogShowing;
+    private EditText mUserAgentCustom;
+ 
+    public UserAgentListPreference(Context context) {
+        super(context);
+        mContext = context;
+    }
+ 
+    public UserAgentListPreference(Context context, AttributeSet attrs) {
+        super(context, attrs);
+        mContext = context;
+    }
+ 
+    @Override
+    protected void onDialogClosed(boolean positiveResult) {
+        super.onDialogClosed(positiveResult);
+        if (positiveResult) {
+            SharedPreferences prefs = PreferenceManager
+                    .getDefaultSharedPreferences(mContext);
+            mUserAgent = prefs
+                    .getString(
+                            MessagingPreferenceActivity.USER_AGENT,
+                            mContext
+                                    .getString(R.string.pref_key_mms_user_agent_default));
+            if (mUserAgent.equals("custom")) {
+                showDialog();
+            }
+        }
+    }
+ 
+    private void showDialog() {
+        LayoutInflater inflater = (LayoutInflater) mContext
+                .getSystemService(Context.LAYOUT_INFLATER_SERVICE);
+ 
+        View v = inflater.inflate(R.layout.custom_user_agent_dialog, null);
+        mUserAgentCustom = (EditText) v.findViewById(R.id.CustomUserAgentEditText);
+ 
+        new AlertDialog.Builder(mContext).setIcon(
+                android.R.drawable.ic_dialog_info).setTitle(
+                R.string.pref_title_mms_user_agent).setView(v)
+                .setOnCancelListener(new OnCancelListener() {
+                    public void onCancel(DialogInterface dialog) {
+                        mDialogShowing = false;
+                    }
+                }).setNegativeButton(android.R.string.cancel,
+                        new DialogInterface.OnClickListener() {
+                            public void onClick(DialogInterface dialog,
+                                    int whichButton) {
+                                mDialogShowing = false;
+                            }
+                        }).setPositiveButton(android.R.string.ok,
+                        new DialogInterface.OnClickListener() {
+                            public void onClick(DialogInterface dialog,
+                                    int whichButton) {
+                                mDialogShowing = false;
+                                SharedPreferences prefs = PreferenceManager
+                                        .getDefaultSharedPreferences(mContext);
+                                SharedPreferences.Editor settings = prefs
+                                        .edit();
+                                settings
+                                        .putString(
+                                                MessagingPreferenceActivity.USER_AGENT_CUSTOM,
+                                                mUserAgentCustom.getText().toString());
+                                settings.commit();
+ 
+                                Toast.makeText(
+                                                mContext,
+                                                mContext.getString(R.string.pref_mms_user_agent_custom_set) + mUserAgentCustom.getText().toString(),
+                                                Toast.LENGTH_LONG).show();
+                            }
+                        }).show();
+        mDialogShowing = true;
+    }
+ 
+    @Override
+    protected void onRestoreInstanceState(Parcelable state) {
+        super.onRestoreInstanceState(state);
+        if (mDialogShowing) {
+            SharedPreferences prefs = PreferenceManager
+                    .getDefaultSharedPreferences(mContext);
+            mUserAgent = prefs
+                    .getString(
+                            MessagingPreferenceActivity.USER_AGENT,
+                            mContext
+                                    .getString(R.string.pref_key_mms_user_agent_default));
+            showDialog();
+        }
+    }
+ 
+    @Override
+    protected View onCreateDialogView() {
+        mDialogShowing = false;
+        return super.onCreateDialogView();
+    }
+ 
+    public void onStartTrackingTouch(SeekBar seekBar) {
+    }
+ 
+    public void onStopTrackingTouch(SeekBar seekBar) {
+    }
+}
+
diff -r -N -u openeclair-android_packages_apps_Mms-4e5c488/src/com/android/mms/transaction/HttpUtils.java openeclair-android_packages_apps_Mms-4e5c488-ua/src/com/android/mms/transaction/HttpUtils.java
--- openeclair-android_packages_apps_Mms-4e5c488/src/com/android/mms/transaction/HttpUtils.java	2010-03-07 21:27:58.000000000 +0000
+++ openeclair-android_packages_apps_Mms-4e5c488-ua/src/com/android/mms/transaction/HttpUtils.java	2010-03-08 15:32:52.219084970 +0000
@@ -31,7 +31,11 @@
 
 import com.android.mms.MmsConfig;
 import com.android.mms.LogTag;
+import com.android.mms.R;
+import com.android.mms.ui.MessagingPreferenceActivity;
 
+import android.preference.PreferenceManager;
+import android.content.SharedPreferences;
 import android.content.Context;
 import android.net.http.AndroidHttpClient;
 import android.telephony.TelephonyManager;
@@ -65,6 +69,8 @@
         HDR_VALUE_ACCEPT_LANGUAGE = getHttpAcceptLanguage();
     }
 
+    private static String mUserAgent;
+
     // Definition for necessary HTTP headers.
     private static final String HDR_KEY_ACCEPT = "Accept";
     private static final String HDR_KEY_ACCEPT_LANGUAGE = "Accept-Language";
@@ -100,6 +106,7 @@
             Log.v(TAG, "httpConnection: params list");
             Log.v(TAG, "\ttoken\t\t= " + token);
             Log.v(TAG, "\turl\t\t= " + url);
+            Log.v(TAG, "\tUser-Agent\t\t=" + mUserAgent);
             Log.v(TAG, "\tmethod\t\t= "
                     + ((method == HTTP_POST_METHOD) ? "POST"
                             : ((method == HTTP_GET_METHOD) ? "GET" : "UNKNOWN")));
@@ -119,7 +126,7 @@
                     hostUrl.getHost(), hostUrl.getPort(),
                     HttpHost.DEFAULT_SCHEME_NAME);
 
-            client = createHttpClient();
+            client = createHttpClient(context);
             HttpRequest req = null;
             switch(method) {
                 case HTTP_POST_METHOD:
@@ -253,9 +260,17 @@
         throw new IOException(exception.getMessage());
     }
 
-    private static AndroidHttpClient createHttpClient() {
-        String userAgent = MmsConfig.getUserAgent();
-        AndroidHttpClient client = AndroidHttpClient.newInstance(userAgent);
+    private static AndroidHttpClient createHttpClient(Context context) {
+
+        // Get Shared Preferences and User Defined User Agent for MMS
+        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
+        mUserAgent = prefs.getString(MessagingPreferenceActivity.USER_AGENT, context.getString(R.string.pref_key_mms_user_agent_default));
+        if (mUserAgent.equals("custom")) {
+            mUserAgent = prefs.getString(MessagingPreferenceActivity.USER_AGENT_CUSTOM, context.getString(R.string.pref_key_mms_user_agent_default));
+        }
+
+        AndroidHttpClient client = AndroidHttpClient.newInstance(mUserAgent);
+
         HttpParams params = client.getParams();
         HttpProtocolParams.setContentCharset(params, "UTF-8");
 
@@ -264,7 +279,7 @@
 
         if (Log.isLoggable(LogTag.TRANSACTION, Log.DEBUG)) {
             Log.d(TAG, "[HttpUtils] createHttpClient w/ socket timeout " + soTimeout + " ms, "
-                    + ", UA=" + userAgent);
+                    + ", UA=" + mUserAgent);
         }
         HttpConnectionParams.setSoTimeout(params, soTimeout);
         return client;
diff -r -N -u openeclair-android_packages_apps_Mms-4e5c488/src/com/android/mms/ui/MessagingPreferenceActivity.java openeclair-android_packages_apps_Mms-4e5c488-ua/src/com/android/mms/ui/MessagingPreferenceActivity.java
--- openeclair-android_packages_apps_Mms-4e5c488/src/com/android/mms/ui/MessagingPreferenceActivity.java	2010-03-07 21:27:58.000000000 +0000
+++ openeclair-android_packages_apps_Mms-4e5c488-ua/src/com/android/mms/ui/MessagingPreferenceActivity.java	2010-03-08 15:33:33.507075556 +0000
@@ -60,6 +60,9 @@
 	public static final String CONVERSATION_FONT_SIZE      = "pref_key_mms_conversation_font_size";
 	public static final String CONVERSATION_HIDE_NAMES = "pref_key_conversation_hide_names";
 	public static final String CONVERSATION_LEFT_RIGHT = "pref_key_conversation_left_right";
+    public static final String USER_AGENT                   = "pref_key_mms_user_agent";
+    public static final String USER_AGENT_CUSTOM            = "pref_key_mms_user_agent_custom";
+
 
     // Menu entries
     private static final int MENU_RESTORE_DEFAULTS    = 1;
